qiskit/visualization/counts_visualization.py

```diff
@@ -46,6 +46,28 @@ def hamming_distance(str1, str2):
 DIST_MEAS = {"hamming": hamming_distance}
 
 
+def _is_deprecated_data_format(data) -> bool:
+    if not isinstance(data, list):
+        data = [data]
+    for dat in data:
+        if isinstance(dat, (QuasiDistribution, ProbDistribution)) or isinstance(
+            next(iter(dat.values())), float
+        ):
+            return True
+    return False
+
+
+@deprecate_arg(
+    "data",
+    deprecation_description=(
+        "Using plot_histogram() ``data`` argument with QuasiDistribution, ProbDistribution, or a "
+        "distribution dictionary"
+    ),
+    since="0.22.0",
+    additional_msg="Instead, use ``plot_distribution()``.",
+    predicate=_is_deprecated_data_format,
+    pending=True,
+)
```

Comment:
This is starting to add non-trivial overhead into any call to the function, since we now loop through all the data now.  It's not important for this visualisation function and doesn't need changing, but I think it's something we should be careful about _in general_.

Reply:
https://github.com/Qiskit/qiskit-terra/issues/9902 should avoid the duplication of being both in the decorator and the function body.

If you'd like, I can prioritize that for next week? Or remove this particular call?

Reply:
We can just leave this call in place for now - the inefficiency isn't going to be an issue for this particular function call.